### ISSUE TITLE
ping input providing false/unreliable statistics with multiple url's

### ISSUE THREAD

**Issue by redrumdk** [influxdata/telegraf#8980]
Hi Guys, 

I have a big concern that the statistics collected by the native method of ping are incorrect. 
The Maximum and Minimum response times are often the same for multiple different url's (ip's)
This makes me think that the averages might also be incorrect, as the min and max used would also reflex the average. 

it looks like the statistics collected are somehow crossed and reused, and hence not correct. 

i have attached a screenshot of the result collected, and color coded the entries i see where the number is the same for the differnt url's 

https://ibb.co/0XStJwk

Debug logs provide no errors or additional information

would you please have a look at it.
thans for all the great work by the way

### Relevant telegraf.conf:

```toml

[[inputs.ping]]
method="native"
urls = ["x.x.x.x",
"x.x.x.x",
"x.x.x.x",
"x.x.x.x", 
"x.x.x.x", 
"x.x.x.x", 
"x.x.x.x", 
"x.x.x.x", 
"x.x.x.x", 
"x.x.x.x",
"x.x.x.x"
] 
count = 10
timeout = 60.0

```

### System info:

Telegraf 17.3, windows 10, windows 2016

### Steps to reproduce:

1. Use the same config as pasted above, and check the database field maximum_response_ms and minimum_response_ms

### Expected behavior:

Individual measurements

### Actual behavior:

a pattern og the same measurements/statistics being reused for individual url's

**Comment by redrumdk** [comment]
@sjwang90
@ssoroka

**Comment by sjwang90** [comment]
@redrumdk Thanks for bringing up the issue. Is this a recent behavior?

**Comment by redrumdk** [comment]
> @redrumdk Thanks for bringing up the issue. Is this a recent behavior?

Yes, since latest release of telegraf 17.3. Ever since the Native method was Fixed.

**Comment by redrumdk** [comment]
same problem in v1.18.1

**Comment by hansenpansen** [comment]
I am experiencing the same with v1.18.1 on Oracle Linux 8.3. Raw capabilities have been enabled in the systemd unit-file for telegraf.

It does not matter if I have a single `[[inputs.ping]]` with multiple elements in `urls`, or multiple `[[inputs.ping]]` sections with a single element in `urls`.

**Comment by redrumdk** [comment]
I am also experiencing random packetloss reported for various ip's with telegraf, while I have been testing with blackbox exporter and prometheus, and ofcourse manually pinging in windows. And they report no packetloss.

Scenatio is:
i have 1 site im pinging which is permanantly down, and my theory is that some of the results from this is being mixed up with the rest of the result. 
As i have been with the other result initially creating the case. 
[https://ibb.co/0XStJwk]

**Comment by toomuchio** [comment]
Also experiencing this issue with Debian Linux, I've had to revert back to exec mode which does not have this issue.

It could be due to conflicting IDs generated for ICMP but that might be a lib issue
https://github.com/go-ping/ping/blob/master/ping.go#L100
https://github.com/go-ping/ping/blob/ff8be3320020b96856d3d48505bcf8926809752e/ping.go#L767

Those don't look completely RNG safe.

**Comment by redrumdk** [comment]
> Also experiencing this issue with Debian Linux, I've had to revert back to exec mode which does not have this issue.
> 
> It could be due to conflicting IDs generated for ICMP but that might be a lib issue
> https://github.com/go-ping/ping/blob/master/ping.go#L100
> https://github.com/go-ping/ping/blob/ff8be3320020b96856d3d48505bcf8926809752e/ping.go#L767
> 
> Those don't look completely RNG safe.

Yeah, im on windows, going back to exec results in another error with that many urls im trying to use.

**Comment by toomuchio** [comment]
What error are you getting @redrumdk ?

For the devs this is a visualization of what native vs exec looks like.
![image](https://user-images.githubusercontent.com/14160086/119613342-43d2b980-be40-11eb-9e02-50448117954d.png)

**Comment by redrumdk** [comment]
> What error are you getting @redrumdk ?
> 
> For the devs this is a visualization of what native vs exec looks like.
> ![image](https://user-images.githubusercontent.com/14160086/119613342-43d2b980-be40-11eb-9e02-50448117954d.png)

For the Native, you can check my screenshot and text in the first post. Im the one opening the case :)

For exec i have another case here. 
https://github.com/influxdata/telegraf/issues/8704

Getting this error in the log, when i have too many IP's for Telegraf to handle 
[inputs.ping] Error in plugin: , exit status 3221225794: X.X.X.X
And then i doesn't collect any measurement for the IP in x.x.x.x

**Comment by srebhan** [comment]
@redrumdk and @toomuchio does this issue still exist on recent versions of Telegraf?

**Comment by redrumdk** [comment]
sure does

**Comment by srebhan** [comment]
@redrumdk I looked into the issue and I think @toomuchio is correct in [his comment](https://github.com/influxdata/telegraf/issues/8980#issuecomment-848478418).

#### Background 
 Telegraf uses one thread per "pinger" so from the OS side of things the only way to distinguish the ICMP packets is the packet ID. The ID in the library is [set to a random 16-bit value](https://github.com/prometheus-community/pro-bing/blob/112c6d152733673e7e7b463bd8a339230536260d/ping.go#L111).

The question now is: How likely is it we do see the same `ID` for two pingers which would lead to the confusion you describe? This problem is known as "Birthday problem" and the probability for `n` hosts to ping is in our case $p = 1 - e^{-\frac{n \cdot (n-1)}{2 \cdot 2^{16}}}$.
While 16-bit seems a lot for 100 endpoints we already get a collision probability of **7.3%**, for 200 we see **26.2%** and for 500 it's **85.1%**!!!

#### Plan

Long story short, the fix is to
1. Set the ID deterministically **across all ping plugins** to **ensure** we don't see collisions.
2. Do not allow more than 65535 IPs in total over all plugins.

I will follow up with a PR, can you please test it @redrumdk?

**Comment by redrumdk** [comment]
if you make a build, i will test

**Comment by srebhan** [comment]
@redrumdk please test the binary in PR #18870, available after CI finished the tests, and let me know if this fixes the issue!?

**Comment by srebhan** [comment]
@redrumdk any chance you test the PR?

**Comment by redrumdk** [comment]
Yes, it's running in production now.
I will be looking for any of the earlier symptoms, thank you very much

**Comment by srebhan** [comment]
@redrumdk any update? I would love to have this in v1.38.5 if possible and it still needs a review...

**Comment by redrumdk** [comment]
It's been running for several days now, and I haven't seen the same pattern
yet.
So far so good 👍🏻

**Comment by srebhan** [comment]
Thanks! Will push the PR to the review pipeline then...
